### PR TITLE
Add mIoU metrics for segmentation model

### PR DESCRIPTION
**Pull request by manhcuong02**
This pull request introduces a new metric, mIoU (mean Intersection over Union), for the segmentation model. The addition of mIoU not only allows the segmentation model to be compared with other segmentation models but also provides additional insights into accuracy for evaluating model performance.

These modifications have been carefully tested to ensure that all other tasks continue to run smoothly. Thank you for considering this contribution

I have read the CLA Document and I hereby sign the CLA

## 🛠️ PR Summary

<sub>Made with ❤️ by [Ultralytics Actions](https://github.com/ultralytics/actions)<sub>

### 🌟 Summary
Enhancements and adjustments added to improve model evaluation on segmentation tasks.

### 📊 Key Changes
- Implemented mean Intersection over Union (mIoU) calculation for segmentation tasks, enhancing the evaluation metrics.
- Refined logging and output information during validation to include new segmentation metrics, notably the mIoU and class-wise mIoU, providing a clearer understanding of model performance.
- Adjusted class results methods to support segmentation-specific metrics, enabling more detailed performance analysis per class.

### 🎯 Purpose & Impact
- **Consistency in Input Handling:** With the standardized input sizes through padding, models can process data more consistently, potentially improving performance.
- **Enhanced Model Evaluation:** The introduction of mIoU for segmentation tasks allows for a more nuanced assessment of model accuracy in distinguishing between different object classes. This could lead to better model optimizations for tasks requiring precise segmentation, such as medical image analysis or autonomous driving.
- **Improved Insights:** By providing detailed metrics, including class-wise mIoU, developers can gain deeper insights into their models' strengths and weaknesses, guiding more targeted improvements and potentially leading to more effective and efficient models.
- **Simplified User Experience:** Adjusting the output during the validation phase to include detailed segmentation metrics directly addresses the needs of developers working on segmentation tasks, simplifying the process of model evaluation and iteration.

These changes collectively aim to enhance the model development and evaluation process for tasks involving segmentation, potentially leading to more accurate and robust models capable of meeting the specific needs of various applications. 📈🔍

### CODE REVIEW

**Comment by manhcuong02** [comment]
I have read the CLA Document and I sign the CLA

**Comment by manhcuong02** [comment]
recheck

File: ultralytics/engine/validator.py
Comment (Burhan-Q) [inline ultralytics/engine/validator.py:20]: Please restore this line

**Comment by glenn-jocher** [comment]
@manhcuong02 certainly! To recheck, you can simply rerun the validation or testing command for your model. For example, if you're validating a YOLOv8 model on a dataset, you might use:

```bash
yolo val model=yolov8n.pt data=your_dataset.yaml
```

Replace `yolov8n.pt` with your model file and `your_dataset.yaml` with your dataset configuration file. This will re-evaluate your model against the dataset. Let me know if you need further assistance! 😊

**Review by Burhan-Q — CHANGES_REQUESTED** [review]
@manhcuong02 thank you for opening a PR. I'm not able to assess the function of your code, as I'm not terribly familiar with the need or purpose, however I'm adding my input with respect to code structure and compatibility. My recommendations are provided to help give your PR the best chance to be accepted.

1. The `mIoU` was not added to the `SegmentMetric.keys`
1. The `mIoU` attribute should be a class property of `SegmentMetrics` as are all other metrics and include a docstring.
1. It's not clear to me there was a need to include indexing with `c` with `self.metrics.class_result()` but I'm not testing the code, I'm just reading it.
1. Using the `.sum()` method (and all `torch.Tensor` methods) is preferred over `torch.sum` as it is faster to execute
1. It will be helpful to see results from your testing of how these changes will impact processing time and hardware resources. Checking run time, system RAM usage, and GPU RAM usage with and without these changes will be helpful to a decision on accepting your proposed changes.

In addition, I suspect that the way this is presently structured does not align with the existing schema (like not using the class properties). You'll want to conform to the existing code schema as best as you can for compatibility and consistency. In its present state, your changes to appear, to me, as though they align.

**Comment by manhcuong02** [comment]
@Burhan-Q 
Thank you for your feedback on this issue. I'd like to respond to some of your questions as follows:
1.  have added mIoU to SegmentMetric.keys as shown below:
image
2. I have also added the mIoU attribute as a class property of SegmentMetrics. I added it as shown below:
image
3. I added "c" to self.metrics.class_result() to serve the validation process. From my understanding, the value of "c" is different from "i", where "c" represents classes with no instances in the ground truth labels. I added "c" to synchronize with the display of data for other parameters within the same class.
4. Thank you for the optimizations you suggested
5. Regarding the time and memory issues, I was fortunate to be able to utilize the functions that you had previously written. I only added a few basic addition and subtraction operations on two matrices with self.nc elements with float32. I also tested it and found no difference compared to the previous version
Below is the image of my experiments on your coco128-seg dataset:
image
image

If I have misunderstood any of your points, please provide feedback so I can make corrections. Thank you sincerely.

File: ultralytics/engine/validator.py
Comment (manhcuong02) [reply ultralytics/engine/validator.py:20]: oh, I'm sorry

**Comment by manhcuong02** [comment]
I have read the CLA Document and I sign the CLA

**Comment by Burhan-Q** [comment]
@manhcuong02 hey, thanks for your notes and make those updates! A few follow-ups from my side:

1. Sorry if you had the `mIoU` in the keys already I hadn't seen that. However, from your screenshot, I'm not sure it's working expected as [this line](https://github.com/ultralytics/ultralytics/blob/f8f62bc649d0554974be34e42725a248b8125cb1/ultralytics/models/yolo/detect/val.py#L178) should mean that the print out for `mIoU` at the end of validation would be a column in the results table.
1. So I meant as a class property as in something like this:

    ```py
        @property
        def mean_iou(self, i):
            """Docstring"""
            return self.seg.mIoU_list[i] # not sure if it would be mIoU or mIoU_list that would go here

            # NOTE self.seg is an instance of the Metric class
    ```
    This would conform to the design of the other metrics for the `SegmentMetrics` class instances. This also means that you'll need to work inside the `SegmentMetrics.seg` attribute, which is an instance of the `Metric` class. This will help your additions use the same interface as the existing code structuring. 

    I'm not certain but it _may_ also mean that your calculations for `miou` may need to be moved around. In addition (I didn't include this in my previous comments), please be certain to include any new class attributes or methods added to the appropriate docstrings.

1. Here I'm not sure if `c` makes sense to use since all others use `i` for indexing, not sure _why_ this is the case, but was only pointing it out. My recommendation would be to test if indexing for your `mIoU` metrics works with `i`. If there are no issues using `i` for indexing, then use that instead, if there are, sharing that would also be helpful. I suspect it _might_ be a difference in data type, `float` vs `int` but again, not sure.

1. You're welcome, I know they would have been the brought up as a requirement.

1. Glad to hear the testing went well. If you're able to share a plot of the system RAM and GPU RAM usage with and without your changes, it would be a great help to have before additional testing is run.

**Comment by Burhan-Q** [comment]
Also it looks like there are some [tests failing](https://github.com/ultralytics/ultralytics/actions/runs/8205389779/job/22442127086?pr=8762#step:6:163), please make sure that your changes will be compatible with all tests (or add/update tests where appropriate). 

<details><summary>Traceback error</summary>
<p>

```
Traceback (most recent call last):

  File "/opt/hostedtoolcache/Python/3.8.18/x64/bin/yolo", line 8, in <module>
    sys.exit(entrypoint())
    
  File "/home/runner/work/ultralytics/ultralytics/ultralytics/cfg/__init__.py", line 568, in entrypoint
    getattr(model, mode)(**overrides)  # default args from model
    
  File "/home/runner/work/ultralytics/ultralytics/ultralytics/engine/model.py", line 513, in val
    validator(model=self.model)
    
  File "/opt/hostedtoolcache/Python/3.8.18/x64/lib/python3.8/site-packages/torch/autograd/grad_mode.py", line 27, in decorate_context
    return func(*args, **kwargs)
    
  File "/home/runner/work/ultralytics/ultralytics/ultralytics/engine/validator.py", line 196, in __call__
    self.update_metrics(preds, batch)
    
  File "/home/runner/work/ultralytics/ultralytics/ultralytics/models/yolo/segment/val.py", line [149](https://github.com/ultralytics/ultralytics/actions/runs/8205389779/job/22442127086?pr=8762#step:6:150), in update_metrics
    self.gt_instances += cls.cpu().long().reshape(-1).bincount(minlength=self.nc)

RuntimeError: output with shape [1] doesn't match the broadcast shape [17]

=============================== warnings summary ===============================
tests/test_python.py::test_export_onnx
  /opt/hostedtoolcache/Python/3.8.18/x64/lib/python3.8/site-packages/torch/tensor.py:587: RuntimeWarning: Iterating over a tensor might cause the trace to be incorrect. Passing a tensor of different shape won't change the number of iterations executed (and might lead to errors or silently give incorrect results).
    warnings.warn('Iterating over a tensor might cause the trace to be incorrect. '

-- Docs: https://docs.pytest.org/en/stable/how-to/capture-warnings.html

---------- coverage: platform linux, python 3.8.18-final-0 -----------
Coverage XML written to file coverage.xml

=========================== short test summary info ============================

FAILED tests/test_cli.py::test_fastsam - subprocess.CalledProcessError: Command '['yolo', 'segment', 'val', 'segment', 'model=/home/runner/work/ultralytics/ultralytics/weights/FastSAM-s.pt', 'data=coco8-seg.yaml', 'imgsz=32']' returned non-zero exit status 1.

======== 1 failed, 68 passed, 8 skipped, 1 warning in 446.02s (0:07:26) ========
Error: Process completed with exit code 1.
```

</p>
</details>

**Comment by manhcuong02** [comment]
@Burhan-Q  Thank you for your feedback. Here are my responses to some of your points:
1. I have added the `mIoU `parameter to the line `pf = "%22s" + "%11i" * 2 + "%11.3g" * len(self.metrics.keys)` as you suggested, and it worked smoothly without any errors. You can view the results in the images I sent earlier on the `coco128-seg dataset`.
2. I understand your point. However, it seems that `mIoU `is only relevant for segmentation tasks. If I add it to the `Metric class` as you suggested, would it be redundant compared to tasks such as classification or detection?
3. I tried using `i` but received unexpected results because it iterated through each class sequentially, including those with no instances in the ground truth labels. For example, if  `class 1` has no instances, then `class 2` would receive the mIoU value of class 1. I noticed that `c`  was used in `self.name[c] of LOGGER.info(pf % (self.names[c], self.seen, self.nt_per_class[c], *self.metrics.class_result(i, c)))` to address this issue. Therefore, I used it to synchronize the parameters of each class. Please refer to the images for the results I obtained with the `coco128-seg` dataset.
4. Thanks
5. I am currently using resources provided by my organization (I have been licensed for this task and also permitted to contribute to this source code), and I am not sure how to access them. I also tried running it on Google Colab and everything seemed to work fine without significant differences. If you have any suggestions regarding this matter, please let me know as there may be some Yolov8 parameters that I am not fully familiar with.

Addition:
In response to your second suggestion within the same comment for ease of tracking, I've included it as point 6.
6. Can you explain the difference between `Tests `and `Benchmarks`? It seems they both run on the same datasets, but tests are showing errors. I have tested on Ubuntu 22.04 and Google Colab with `coco128-seg`, and everything seems to run without issues. I am curious about this discrepancy. Thank you very much

**Comment by Burhan-Q** [comment]
1. I see, I missed the column header, thank you for point that out! I was confused by the second image, seeing the `class_mIoU` not printed as a column.
1. I believe you can initialize it with an empty `list` or `None` value. Other attributes are implemented in a similar manner and you can see that an `Metric` class instance used for the `self.boxes` and `self.seg` attributes of the `SegmentMetrics` class.
    https://github.com/ultralytics/ultralytics/blob/d24ead447eb5747e9d434d7e81b2ca20f2f31cb9/ultralytics/utils/metrics.py#L928-L929
You may choose to leave it as-is, but I suspect (I don't know for certain) that the final reviewer will make the same request which is why I make the recommendation. Let me know if you prefer to keep it as-is and I will remove my `change-request` review.
1. Understood, I will take you on your word and let the final reviewer make the decision here.
1. 👍 
1. For measuring system memory usage, you can use a utility such as `memory_profiler` to track changes in system RAM allocations between changed/unchanged versions. As reference for something similar in a recent PR, see this comment for [GPU memory profiling](https://github.com/ultralytics/ultralytics/pull/8645#issuecomment-1984510992) and this one for [system memory profiling](https://github.com/ultralytics/ultralytics/pull/8645#issuecomment-1983613251).
1. Benchmark CI tests ensure that new changes don't hurt model performance. If a model performs under the set threshold, it will fail. The Tests CI check to make sure that all classes, methods, functions, tasks, and modes work as expected. When a specific PR fails, that PR either introduces a breaking change in the modifications made or with the expected behavior. This is what would need to be corrected prior to final review of your PR. The first thing to do is update your PR branch with the latest `main` branch from Ultralytics to ensure all tests complete successfully, if they don't then you'll need to track down the source of the test failure.

**Comment by Burhan-Q** [comment]
I updated your branch with the latest of the Ultralytics `main` branch and this will start running checks. One last note, you also need to make sure to sign the CLA, see [this comment](https://github.com/ultralytics/ultralytics/pull/8762#issuecomment-1984004457) for more information.

**Comment by Burhan-Q** [comment]
The test that's failing can be run locally using `pytest tests/test_cli.py::test_fastsam` I ran with `ultralytics 8.1.26` locally and it was able to pass without any issues. 
image
You'll need to assure the changes made are able to pass this test for your PR to be accepted.

**Comment by manhcuong02** [comment]
@Burhan-Q  Thank you for your feedback. Here are my responses to some of the issues you pointed out:
1. and 3. Thanks
2. Yes, I understand, and I have made the necessary modifications within the `Metric()` class.
5. I will try it out. Thank you.
6. I ran` !pytest test/test_cli.py::test_fastsam` as you suggested. However, I encountered an issue with your `FastSAM model` on the `coco8-seg` dataset. When I tried to display `cls` and `self.nc`, I noticed that `self.nc = 1` while `cls = 16` and the number of classes in the coco dataset is 80 (`nc = 80`). I wonder if there might be a problem with the FastSAM segmentation model because when I changed it to `yolov8s-seg.pt` and reran `!pytest test/test_cli.py::test_fastsam`, everything worked perfectly fine without any errors. I also found in your documentation that FastSAM does not support the validation process as shown in the image below: 
image
Please let me know your opinion on this matter. Am I misunderstanding something? I sincerely appreciate your input.

**Comment by Burhan-Q** [comment]
@manhcuong02 you're correct about the documentation, and there's definitely something strange happening. I looked closer at  `test_fastsam` and found that it's running two CLI commands at the start:

```py
   # These are the default args
    task="segment"
    model="FastSAM-s.pt"
    data="coco8.seg.yaml"
    
    run(f"yolo segment val {task} model={model} data={data} imgsz=32")
    run(f"yolo segment predict model={model} source={source} imgsz=32 save save_crop save_txt")
```

With `ultralytics==8.1.27` I ran `yolo segment val model=FastSAM-s.pt data=coco8-seg.yaml imgsz=32` and this completed without any errors. I also went a head and tested this in python directly, also with no errors

image

I think that the error may stem from the `Metric` component of the output, as when I attempt to access certain attributes from the `results` object, I get `AttributeError: 'Metric' object has no attribute...` which is unsurprising for since this is not supported for FastSAM. I suspect that if you comment the line `run(f"yolo segment val {task} model={model} data={data} imgsz=32")` and run again, it might not fail?

**Comment by Burhan-Q** [comment]
@glenn-jocher any reason that `run(f"yolo segment val {task} model={model} data={data} imgsz=32")` is needed for testing in `tests/test_cli.py::test_fastsam` since it's not supported per the docs? Perhaps this is used as a "generic" test for the `.val` method or has some other purpose? If not, should this line be commented or removed from testing?

**Comment by manhcuong02** [comment]
@Burhan-Q Thank you for your response. Indeed, when running `yolo segment val model=FastSAM-s.pt data=coco8-seg.yaml imgsz=32`, no errors occur. However, you can see that metrics such as mAP, Precision, and Recall all return 0 because FastSAM-s.pt incorrectly predicts objects in the image, as its actual class count is `only one` while the `ground truth can have up to 80 classes` from the coco dataset. As for commenting or removing it, I am currently unsure about this issue. What do you think about changing the test set for yolov8s-seg.pt and retesting?

**Comment by Burhan-Q** [comment]
@manhcuong02 since per the docs I don't believe `FastSAM` is intended to support `val` mode, I'm thinking that it may be viable to remove this from our testing, but that's something Glenn will have to make the final call on, which is why I mentioned him in my last [comment](https://github.com/ultralytics/ultralytics/pull/8762#issuecomment-1991873482).

**Comment by manhcuong02** [comment]
I have read the CLA Document and I sign the CLA

**Comment by manhcuong02** [comment]
> @glenn-jocher any reason that `run(f"yolo segment val {task} model={model} data={data} imgsz=32")` is needed for testing in `tests/test_cli.py::test_fastsam` since it's not supported per the docs? Perhaps this is used as a "generic" test for the `.val` method or has some other purpose? If not, should this line be commented or removed from testing?

@glenn-jocher Could you please provide your opinion on this issue? We are discussing the problem with the `FastSAM` model in `!pytest test/test_cli.py::test_fastsam`, which does not support the validation process and only has `1 class` while the `coco8` dataset contains `80 classes`. Thanks!

**Comment by manhcuong02** [comment]
> I updated your branch with the latest of the Ultralytics `main` branch and this will start running checks. One last note, you also need to make sure to sign the CLA, see [this comment](https://github.com/ultralytics/ultralytics/pull/8762#issuecomment-1984004457) for more information.

@Burhan-Q I have also looked into this issue. It seems that I have added my email and successfully signed your CLA. However, it appears that the first time I committed, I did not add my email to my account. Therefore, the error "Your Name seems not to be a GitHub user" still persists, and I have not successfully signed the CLA yet. Do you have any advice for me on this issue? Thank you!

**Comment by Burhan-Q** [comment]
@manhcuong02 you'll need to add a comment with the [text shown here](https://github.com/ultralytics/ultralytics/pull/8762#issuecomment-1984004457). Then you can write another comment after that with only the word `recheck` to initialize the CLA Assistant CI checker.

**Comment by manhcuong02** [comment]
@Burhan-Q Thank you for the advice. I have tried before but it didn't work. The "Your Name" account is still there (it was like that the first time I committed and it is still the same now). I don't know how to remove it. I think maybe I will have to create a new pull request to fix this. Do you have any thoughts on this?

Addition:
We also haven't received feedback on the `Tests checks` from @glenn-jocher regarding this matter. If we create a new pull request, it still won't be able to check successfully.

**Comment by Burhan-Q** [comment]
recheck

**Pull request closed without merging**
